File: solutions/javascript/all-your-base/1/all-your-base.js
//
// This is only a SKELETON file for the 'All Your Base' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const convert = (digits, inputBase, outputBase) => {
  if (inputBase < 2) throw new Error('Wrong input base');
  if (outputBase < 2) throw new Error('Wrong output base');

  if (
    digits.length === 0 ||
    (digits.length > 1 && digits[0] === 0) ||
    digits.some(d => d < 0 || d >= inputBase)
  ) {
    throw new Error('Input has wrong format');
  }

 
  let decimal = 0;
  for (let d of digits) {
    decimal = decimal * inputBase + d;
  }


  if (decimal === 0) return [0];

  const outputDigits = [];
  while (decimal > 0) {
    outputDigits.unshift(decimal % outputBase);
    decimal = Math.floor(decimal / outputBase);
  }

  return outputDigits;
};

 

File: solutions/python/all-your-base/1/all_your_base.py
"""Convert a number from one base to another."""

def rebase(input_base, digits, output_base):
    """Rebase digits from input_base to output_base."""

    if input_base < 2:
        raise ValueError("input base must be >= 2")

    if output_base < 2:
        raise ValueError("output base must be >= 2")

    
    if not digits:
        return [0]

    for d in digits:
        if d < 0 or d >= input_base:
            raise ValueError("all digits must satisfy 0 <= d < input base")

 
    decimal_value = 0
    for d in digits:
        decimal_value = decimal_value * input_base + d

    
    if decimal_value == 0:
        return [0]

   
    output_digits = []
    while decimal_value > 0:
        output_digits.append(decimal_value % output_base)
        decimal_value //= output_base

    output_digits.reverse()
    return output_digits
